content/en/blog/_posts/2020-05-20-my-exciting-journey-into-kubernetes-history.md

```diff
@@ -0,0 +1,655 @@
+project.
+
+Parameters like `min_df`, `max_df` and the n-gram range can be seen as some of
+our hyperparameters. Those parameters have to be optimized in a dedicated step
+after the machine learning model has been built. This step is called
+hyperparameter tuning and basically means that we train multiple times with
+different parameters and compare the accuracy of the model. Afterwards, we choose
+the parameters with the best accuracy.
+
+[60]: https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html
+[61]: https://en.wikipedia.org/wiki/Bag-of-words_model
+
+During the training, the vectorizer will produce a `data/features.json` which
+contains the whole vocabulary. This gives us a good understanding of how such a
+vocabulary may look like:
+
+```json
+[
+  …
+  "hostname",
+  "hostname address",
+  "hostname and",
+  "hostname as",
+  "hostname being",
+  "hostname bug",
+  …
+]
+```
+
+This produces round about 50,000 entries in the overall bag-of-words, which is
+pretty much. Previous analyses between different data sets showed that it is
+simply not necessary to take so many features into account. Some general data
+sets state that an overall vocabulary of 20,000 is enough and higher amounts do
+not influence the accuracy any more. To do so we can use the [SelectKBest][62]
+feature selector to strip down the vocabulary to only choose the top features.
+Anyway, I still decided to stick to the top 50,000 to not negatively influence
+the model accuracy. We have a relatively low amount of data (appr. 7,000
+samples) and a low number of words per sample (~15) which already made me wonder
+if we have enough data at all.
+
+[62]: https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectKBest.html
+
+The vectorizer is not only able to create our bag-of-words, but it is also able to
+encode the features in [term frequency–inverse document frequency (tf-idf)][63]
+format. That is where the vectorizer got its name, whereas the output of that
+encoding is something the machine learning model can directly consume. All the
+details of the vectorization process can be found in the [source code][64].
+
+[63]: https://en.wikipedia.org/wiki/Tf%e2%80%93idf
+[64]: https://github.com/kubernetes-analysis/kubernetes-analysis/blob/f419ff4a3462bafc0cb067aa6973dc7280409699/src/nlp.py#L193-L235
+
+## Creating the Multi-Layer Perceptron (MLP) Model
+
+I decided to choose a simple MLP based model which is built with the help of the
+popular [TensorFlow][70] framework. Because we do not have that much input data,
+we just use two hidden layers, so that the model basically looks like this:
+
+![model](/images/blog/2020-05-20-my-exciting-journey-into-kubernetes-history/model.png)
+
+[70]: https://www.tensorflow.org/api_docs/python/tf/keras?utm_source=DevSite&utm_campaign=Text-Class-Guide&utm_medium=referral&utm_content=tensorflow&utm_term=keras
```

Comment:
nit: you could remove the tracking link here, it might be misleading for analytics.

Reply:
Yes, I removed that part of the link as well.